### ISSUE TITLE
Hashed Containers (or a Better Explanation why not)

### ISSUE THREAD

**Issue by notevenodd** [glouw/ctl#9]
The README states that CTL will not have unordered_map and unordered_set. The rationale for this is quite succinct: "_ordered containers are preferred, even at the cost of performance_".

If CTL won't offer hashed containers, could it at least extend the explanation why red/black containers should be always preferred. It's not obvious why it would always be true. Especially, as:
1) hash tables are a popular datastructure in C
2) it's not just about popularity, hash containers retrieval is algorithmically faster (O(1) average) which really can matter.

**Comment by glouw** [comment]
Yes, I got quite a bit of interest on hackernews for unordered maps, so I started my implementation of `std::unordered_map` on the staging branch (see `ust.h`): https://github.com/glouw/ctl/blob/staging/ctl/ust.h

**Comment by glouw** [comment]
It should be ready in a couple weeks

**Comment by rurban** [comment]
But very buggy still. And chained, not open addressing. You shalt not make the sand mistake as the STL

Rehash, each are broken, insert needs to check the load_factor

**Comment by glouw** [comment]
@rurban, considering it's the staging branch, I wouldn't consider it production worthy. As for linked list chained hash tables, the main benefit is that pointers to unordered set elements remain valid for the lifetime of the unordered set, which, in practically, is more important than speed. I wouldn't consider this a mistake in the STL.

**Comment by glouw** [comment]
I've back ported `std::unordered_set`, named `ust.h`. It's 2.5x faster than `set.h`. Here are it's performance characteristics:

![image](https://user-images.githubusercontent.com/7000616/104406111-66a8a500-5513-11eb-8625-ca8632eddae8.png)

![image](https://user-images.githubusercontent.com/7000616/104406097-5bee1000-5513-11eb-93c0-bc487f767093.png)

**Comment by glouw** [comment]
Example hashing of strings (with a relatively uninteresting hash function):

```
#include <stdio.h>
#include <str.h>

#define T str
#include <ust.h>

size_t
str_hash(str* s)
{
    size_t hash = 5381;
    int c;
    char* temp = s->value;
    while(c = *temp++)
        hash = ((hash << 5) + hash) + c;
    return hash;
}

int
str_equal(str* a, str* b)
{
    return str_key_compare(a, b) == 0;
}

int
main(void)
{
    ust_str strs = ust_str_init(str_hash, str_equal);
    ust_str_insert(&strs, str_init("this"));
    ust_str_insert(&strs, str_init("is"));
    ust_str_insert(&strs, str_init("a"));
    ust_str_insert(&strs, str_init("test"));
    ust_str_insert(&strs, str_init("for"));
    ust_str_insert(&strs, str_init("unordered"));
    ust_str_insert(&strs, str_init("set"));
    ust_str_insert(&strs, str_init("for"));
    ust_str_insert(&strs, str_init("the"));
    ust_str_insert(&strs, str_init("C"));
    ust_str_insert(&strs, str_init("Standard"));
    ust_str_insert(&strs, str_init("Template"));
    ust_str_insert(&strs, str_init("Library"));
    ust_str_insert(&strs, str_init("(CTL)"));
    ust_str_insert(&strs, str_init("..."));
    foreach(ust_str, &strs, it)
        puts(str_c_str(it.ref));
    for(size_t i = 0; i < strs.bucket_count; i++)
        printf("%2d : %2lu\n", i, ust_str_bucket_size(&strs, i));
    ust_str_free(&strs);
}
```
